File: typescript/test/Wall/wallService.test.ts
import {WallService} from '../../src/Wall/WallService';
import {User} from '../../src/User/User';
import {UserNotLoggedInError} from '../../src/Error/UserNotLoggedInError';
import {UsersAreNotFriendsError} from '../../src/Error/UsersAreNotFriendsError';
import {Brick} from '../../src/Wall/Brick';
import {StubClock} from '../Wrapper/StubClock';
import {FakeWallDAO} from './FakeWallDAO';
import {DummyWallDAO} from "./DummyWallDAO";
import {DummyClock} from "../Wrapper/DummyClock";
import {SpyWallDAO} from "./SpyWallDAO";

const GUEST = undefined;
const REGISTERED_USER = new User();

describe('Wall Service test', () => {

    it('should throw an error when user is not logged in', () => {
        const wallService = new WallService(new DummyWallDAO(), new DummyClock());

        expect(() =>
            wallService.anotherBrickInTheWall(new User(), '', GUEST)
        ).toThrow(UserNotLoggedInError);
    });

    it('should throw an error when user are not friend with', () => {
        const wallService = new WallService(new DummyWallDAO(), new DummyClock());

        expect(() =>
            wallService.anotherBrickInTheWall(new User(), '', REGISTERED_USER)
        ).toThrow(UsersAreNotFriendsError);
    });

    it('should append a new brick to friend wall', () => {
        const stubClock = new StubClock();
        let spyWallDAO = new SpyWallDAO();
        const wallService = new WallService(spyWallDAO, stubClock);
        let user = new User();
        user.addFriend(REGISTERED_USER);

        wallService.anotherBrickInTheWall(user, '', REGISTERED_USER);

        expect(spyWallDAO.userParameter).toEqual(user);
        expect(spyWallDAO.brickParameter).toEqual(new Brick("", stubClock.now()))
    });

    it('appended brick should return the updated wall', () => {
        const stubClock = new StubClock();
        const wallService = new WallService(new FakeWallDAO(), stubClock);

        let user = new User();
        user.addFriend(REGISTERED_USER);

        let wall = wallService.anotherBrickInTheWall(user, '', REGISTERED_USER);

        expect(wall.length).toEqual(1);
        expect(wall[0]).toEqual(new Brick('', stubClock.now()));
    });
});

File: dotNet/test/Wall/WallServiceTest.cs
using src.Exception;
using src.Wall;
using src.Wrapper;
using test.Wrapper;

namespace test.Wall;

public class Tests
{
    private static readonly src.User.User? Guest = null;
    private static readonly src.User.User RegisteredUser = new();

    [Test]
    public void should_Throw_An_Exception_When_User_Is_Not_Logged_In()
    {
        WallService wallService = new WallService(new DummyWallDAO(), new DummyClock());

        Assert.Throws<UserNotLoggedInException>(() =>
            wallService.AnotherBrickInTheWall(new src.User.User(), "", Guest));
    }

    [Test]
    public void should_Throw_An_Exception_When_User_Are_Not_Friend_With()
    {
        WallService wallService = new WallService(new DummyWallDAO(), new DummyClock());

        Assert.Throws<UsersAreNotFriendsException>(() =>
            wallService.AnotherBrickInTheWall(new src.User.User(), "", RegisteredUser));
    }

    [Test]
    public void should_Append_New_Brick_To_Friend_Wall()
    {
        SpyWallDAO spyWallDao = new SpyWallDAO();
        StubClock stubClock = new StubClock();
        WallService wallService = new WallService(spyWallDao, stubClock);
        src.User.User user = new src.User.User();
        user.AddFriend(RegisteredUser);

        wallService.AnotherBrickInTheWall(user, "", RegisteredUser);

        Assert.That(user, Is.EqualTo(spyWallDao.UserParameter));
        Assert.That(new Brick("", stubClock.Now()), Is.EqualTo(spyWallDao.BrickParameter));
    }

    [Test]
    public void appended_Brick_Should_Return_The_Updated_Wall()
    {
        FakeWallDAO fakeWallDao = new FakeWallDAO();
        StubClock stubClock = new StubClock();
        WallService wallService = new WallService(fakeWallDao, stubClock);
        src.User.User user = new src.User.User();
        user.AddFriend(RegisteredUser);

        List<Brick> wall = wallService.AnotherBrickInTheWall(user, "", RegisteredUser);

        Assert.That(wall.Count, Is.EqualTo(1));
        Assert.That(wall[0], Is.EqualTo(new Brick("", stubClock.Now())));
    }
}
